add mirror args to mirror create command

opadron opened the pull request:
Fixes #17531.

Commit:
add mirror args to mirror create command

Commit:
update command completion

scottwittenburg commented on lib/spack/spack/mirror.py:
```diff
@@ -100,6 +100,65 @@ def from_dict(d, name=None):
+
+        if ((args.directory and args.mirror_name) or
+                (args.mirror_name and args.mirror_url) or
+                (args.mirror_url and args.directory)):
```
Are you aware of argparse method `add_mutually_exclusive_group`?  There are other commands that make use of it, it might be useful here and easier to read/understand.

scottwittenburg replied:
I see you've used the built argparse approach to make sure only one of those options is provided, thanks.  Have you kept this check because of how the buildcache cmd arg handling doesn't make use of the same approach?

opadron replied:
I kept this as an extra level of sanity checking.  I figure that if new call sites are introduced, they may not enforce the same constraints when constructing `args`, so I decided not to assume as much...

...although I should probably update the message in the `ValueError` below since we're now one-step removed from CLI arguments.

scottwittenburg reviewed (changes requested):
This looks good to me, except for the one thing I pointed out.  It helps to unify some similar actions between mirrors and buildcaches, and consolidates some code that implements it.  Thanks @opadron!

Commit:
use add_mutually_exclusive_group()

scottwittenburg commented:
@opadron I just noticed the coverage for the patch came back under 50%, any chance you can add a test to bump that?

robo-wylder commented:
I tried it with a command like this and it breaks in the `MirrorCache` class with the following error. I'm not sure this will really be possible without updating the cache class as well, unless you had a different use case in mind.

```shell
spack mirror create --mirror-url s3://my-spack-mirror/ --dependencies boost
```

```
  File "~/spack/lib/spack/spack/mirror.py", line 488, in create
    'MirrorCaches only work with file:// URLs')
spack.error.SpackError: MirrorCaches only work with file:// URLs
```

Commit:
make MirrorCache class work with remote mirrors

Commit:
fix tests

Commit:
fix style

opadron commented:
@robo-wylder you're right, it does look like the `MirrorCache` class needed to be updated to be generic and allow (where possible) non-local mirror URLs.  I've made an attempt to perform this update in the latest few commits.  Please try them again and let me know if they work for you.

@scheibelp @Paul-Ferrell please take a look at the latest commits I've added.  They touch the `MirrorCache` class and related code.

opadron commented:
@scottwittenburg BTW I haven't forgotten about the patch coverage.  I'll get those numbers up once I can get confirmation that the changes work for other users.

scottwittenburg commented:
> @scottwittenburg BTW I haven't forgotten about the patch coverage. I'll get those numbers up once I can get confirmation that the changes work for other users.

Sounds good, thanks @opadron!

opadron commented:
Tagging as "dont-merge-yet".  There's still some testing work I'd like to add before merging.

robo-wylder commented:
@opadron 

Tried out the patch again with the same commands I listed above and got a different error. It looks like there's a point where the cache attempts to validate that the URL of the new cached source is of `file://` type.

```
Traceback (most recent call last):
  File "lib/spack/spack/mirror.py", line 562, in _add_single_spec
    pkg_stage.cache_mirror(mirror, mirror_stats)
  File "lib/spack/spack/util/pattern.py", line 68, in getter
    getattr(item, self.name)(*args, **kwargs)
  File "lib/spack/spack/stage.py", line 539, in cache_mirror
    self.fetcher, self.mirror_paths.storage_path)
  File "lib/spack/spack/caches.py", line 68, in store
    'Cannot store sources into non-remote (file://) URLs')
spack.error.SpackError: Cannot store sources into non-remote (file://) URLs
```

It looks like the check you added in `MirrorCache.store()` for whether the file is a local path or not needs to be removed or the file is never pushed to the remote mirror. There's also an adjacent bug in `botocore` where any path that starts with a `/` or is an empty path cannot be used with the `get_object` method so I think we need to patch `parse()` in `lib/spack/spack/util/url.py` so that it lstrips all slashes except in the case that the entire path is just `/`. Example below for clarity of what I mean. I think this is probably being hit elsewhere but not being noticed, so it may fix a few other issues as well.

```
if scheme == 's3' and path != '/':
    path = path.lstrip('/')
```

In my local testing with the `store()` and `parse()` changes it appears to work both uploading and detecting when the remote files exist and not re-uploading them. Thanks again for the speedy effort in getting this working!

Commit:
fix mirror regression test

robo-wylder commented on lib/spack/spack/caches.py:
```diff
@@ -51,24 +52,35 @@ def _fetch_cache():
         # Note this will archive package sources even if they would not
         # normally be cached (e.g. the current tip of an hg/git branch)
-        dst = os.path.join(self.root, relative_dest)
+        dst = spack.util.url.local_file_path(
+            spack.util.url.join(self.root, relative_dest))
+
+        if not dst:
+            raise spack.error.SpackError(
+                'Cannot store sources into non-remote (file://) URLs')
```
I left a general comment about this but we have to allow the fetcher to write here or the data is never pushed to remote mirrors on creation. Is there a reason you are forcing this to be a local file path? The fetcher contains the logic for how to archive the cached file, I think including any directories that need to exist first.

Commit:
Revert "fix tests"

This reverts commit dceab8ef604190113bdaeb62641a5ca94409ebd6.

Commit:
fix mirror cache not handling remote urls correctly

opadron commented:
@robo-wylder thanks for that catch!  I realized that caching code needed more changes.  Please try again with the latest change. (rebased, `HEAD~1`, since the current `HEAD` is a half-baked commit I'm still iterating on).

As for the `botocore` bug, that should already be handled by the `s3_handler` for `urllib`. (see `spack.s3_handler._s3_open()`).  I think keeping implementation details for working around these kind of idiosyncrasies out of the general `parse()` method should be preferred, but I'll be happy to revisit this question if you continue to run into problems.

Thanks for your help!

robo-wylder commented:
@opadron that is a fair point about the `parse()` method, however, at the very least we must patch the `url_exists()` method in `web.py` in order for the remote mirror check to function properly. It looks like there's a check in the `_s3_open()` method in `s3_handler.py` that contains similar logic as you pointed out. The same logic I proposed above should be sufficient there as well, specifically including the case for when the path is only `/` or it will think buckets do not exist.

I have patched the `parse()` method in my local fork of Spack and if I undo that a number of things seem to fail, but they are mostly related to the `url_exists` method failing when the path starts with a `/`.

robo-wylder commented:
A quick test of this shows it now uploads correctly, however, it does require the patch to `url_exists()` to recognize that the files exist already. Great work @opadron !

opadron commented:
@robo-wylder I just rebased again with another commit fixing `url_exists()`.  Please try it again.

robo-wylder commented on lib/spack/spack/util/web.py:
```diff
@@ -212,7 +212,11 @@ def url_exists(url):
         from botocore.exceptions import ClientError
         try:
-            s3.get_object(Bucket=url.netloc, Key=url.path)
+            key = url.path
+            if key.startswith('/'):
+                key = key[1:]
```
You also need to check if the path is only `/`, otherwise checks for just a bucket will fail (e.g. `s3://my-bucket` will fail but `s3://my-bucket/my-file` will not) because `get_object` expects the `Key` parameters to be a string of length 1 or greater.

Additionally, I'd recommend using `key.lstrip('/')` instead of `key[1:]` so that if there are multiple leading slashes they are removed appropriately.

opadron replied:
The case of `scheme == 's3' and path == '/'` is curious.  We do need this to be a valid case for listing (show me everything in the top level of my bucket), but I probably need to update `url_exists()` to handle this specific case (I'd interpret it as asking "does the bucket exist, at all?"), and I don't think this case even makes sense for fetching.  In that case I think Spack should just raise an exception without even trying.

Good point on the use of `lstrip()`!

opadron replied:
Actually, I may have just talked myself out of my interpretation for `url_exists()`, since I imagine that for most purposes, a generic resource "exists" if and only if I can "fetch" it.  I even rely on this assumption as a fallback (just try to fetch, and assume that any successful result implies existence).

`url_exists()` for the above case should probably just return `False`, since fetching a bucket is undefined.

(I realize this isn't directly related to getting your use case working, again, but I figure this is probably a good spot to sort this out.)

robo-wylder replied:
There's another method you could use called `head_bucket`, but it does technically require the `ListBucket` permission, but that's probably not too much to ask. This method `url_exists()` should definitely be able to check if a bucket exists, but you could modify the logic to call `head_bucket` if the `url.path` is just `/` instead of trying to fetch it. I think the fetch is lazy when it's the whole bucket anyway.

robo-wylder replied:
Just tested this locally and it seems to work:
```python
if url.path == '/':
  s3.head_bucket(Bucket=url.netloc)
else:
  s3.get_object(Bucket=url.netloc, Key=url.path.lstrip('/'))
```

Edit: You'll need to check for a different error code when it doesn't exist.

opadron replied:
I think I'll need to do something like what you're suggesting, after all.  It's occurred to me that fetchable implies exists, but not vice-a-versa.  File system directories are another example and I'm pretty sure we rely on `url_exists()` to return `True` in this case.

So exists implies *either* fetchable or ... "listable"?

Man, what a rabbit hole.  Thanks for humoring me! 😃

Commit:
handle url_exists("") == True when url_exists("/index.html")

Commit:
Stop checking "path" -> "path/index.html" for S3

Commit:
do head_bucket() for url_exists("s3://bucket")

opadron commented:
@robo-wylder give these changes a spin.

I also took the liberty of cutting out the whole section of code assuming that `<url>/index.html` could act as a substitute for `<url>`, since that's really only true for certain web servers and we shouldn't be relying on that behavior anywhere in the code.

In any case, it would be reassuring to know that your use cases aren't affected.

Commit:
fix tests

robo-wylder commented on lib/spack/spack/caches.py:
```diff
@@ -51,24 +55,43 @@ def _fetch_cache():
+            mkdirp(os.path.dirname(local_dst))
+            fetcher.archive(local_dst)
+        else:
+            # In the case of non-local destination, fetch to tmp dir, upload to
+            # remote destination, and then delete tmp dir
+            tmpdir = tempfile.mkdtemp(prefix='spack-mirror-cache-store')
+            try:
+                local_dst = os.path.join(tmpdir, relative_dest)
+                mkdirp(os.path.dirname(local_dst))
+                fetcher.archive(local_dst)
+
+                spack.util.web.push_to_url(local_dst, dst, keep_original=False)
+            finally:
+                shutil.rmtree(tmpdir, ignore_errors=True)
```
I'm not sure this is necessary, the `fetcher.archive()` call itself seems to do the right thing without having to handle remote cases specially. Local directories get created automatically for local paths and for remote paths the fetcher knows how to push the file that already exists locally to the url. Look at the `archive()` method for the `URLFetchStrategy` and the implementation of `push_to_url()` in `web.py`.

I think the issue before was that this always forced a local path instead of just attempting to store the file in `dst` directly. Did the tests fail without the extra logic to localize the path?

opadron replied:
I'm pretty sure the tests did fail, though I should double check it, again.  I'm a bit out of my element since I didn't introduce the `MirrorCache` class and it's not very clear to me exactly what its purpose is.

My current understanding is that it is an additional abstraction layer that is supposed to cache downloads from remote mirrors, yet it was clearly not written to handle non-local URLs (or tested with any), so I'm really confused as to what the original intent was.  I'd imagine a caching layer would somehow map non-local mirror URLs to some root directory under the Spack cache, and serve subsequent fetch requests out of that cache, but I don't see that logic anywhere.

I'll take another look and see if I can't make two cents of what this class does, and what needs to change to get it to do "the right thing" with remote URLs.

EDIT: Taking another look, I seem to recall that the tests passed, but you were observing that some files where never actually getting pushed to your mirror.

robo-wylder replied:
@scheibelp it looks like you were the one who introduced this class, would you mind enlightening us on the intent of it?

Commit:
increase coverage

Commit:
Revert "fix mirror cache not handling remote urls correctly"

This reverts commit fd244be082708fc301edefa5473013e9c01f1ed9.

opadron commented:
I've rebased and revived this PR.  The changes here are needed to resolve #25765.

@scottwittenburg @robo-wylder: would you be willing to take another pass at this and provide any feedback?

@scheibelp: See above for context.  We were stuck trying to understand what the `MirrorCache` class was for and how it was supposed to work.  I think we need to understand this because the changes in this PR affect this class.  We could really use your help.

EDIT TO ADD MORE CONTEXT:

This PR was started with the intention of adding the different "mirror args" to the `mirror create` command.  By "mirror args", I'm referring to the `--mirror-name`, `--mirror-url` and `--directory` arguments (the three different ways to specify a mirror).  It did this by joining these common arguments into an argparser mutually exclusive group and refactoring the validation logic into a centralized location.  With these changes, different subcommands, like `buildcache create` (the first subcommand that used these arguments) and `mirror create` (the original target of this PR) can accept "mirror args" with minimal duplication of code and effort.  Also, other commands that might need to accept a mirror (like `gpg publish`, the subject of #25765), will be able to do the same.

@robo-wylder reported above that these changes had broken the `MirrorCache` class, a new class that was introduced by @scheibelp since the original `buildcache create` command was introduced, and the conversations above is mostly me trying to update the `MirrorCache` class to work with these changes.  Not having introduced the class, I was unable to comprehend its purpose or how it's supposed to work, and without this understanding I was stuck without a way forward.  @scheibelp was asked to help, but I'm guessing he was probably just really busy and we all eventually forgot about this PR.  Until @ashermancinelli reports #25765, eventually leading me to [this](https://github.com/spack/spack/blob/develop/lib/spack/spack/cmd/gpg.py#L203) note I left for myself, and bringing me back to this issue.

I've rebased the branch as of this morning, and that's where we are with this PR.  I still need help (ideally from @scheibelp, or someone who can explain to me the `MirrorCache` class), and given that more than a year has gone by since this morning's rebase, it probably makes sense to review this PR de novo.

tgamblin commented:
@scheibelp can you look at this one again?

Commit:
more style fixes

scheibelp commented:
> We were stuck trying to understand what the MirrorCache class was for and how it was supposed to work.

At the moment, `MirrorCache` manages the root of the mirror directory and replicates the API of `FsCache.store` to place all needed resources there (as requested by `spack.mirror._add_single_spec`); it has a different implementation of `store` that is willing to store anything (including things that wouldn't make sense to cache/reuse implicitly like the tips of SCM branches). It also contains the logic for creating alternative human-readable symlinks to archived resources. 

`MirrorCache` was added in https://github.com/spack/spack/pull/8993 but has simplified since then: the global state has been removed and the `MirrorCache` is created in `spack.mirror.create`; it is now just passed through `mirror._add_single_spec -> Stage.cache_mirror` and isn't intended to do anything beyond what those functions ask of it.

Like the `FsCache`, it depends on `FetchStrategy.archive` to handle the details of actually storing the resource in the specified destination, and I don't see anything immediately wrong with the changes to `MirrorCache` here. Could you mention the errors you are getting?

opadron commented:
In that case, I'll defer back to @robo-wylder.

@robo-wylder: could you give this another review and let us know if you still see any potential issues?

In the meantime, I'll iterate on the CI.

Commit:
fix style

opadron commented on lib/spack/spack/caches.py:
```diff
@@ -59,24 +60,31 @@ def _fetch_cache():
 class MirrorCache(object):
     def __init__(self, root, skip_unstable_versions):
-        self.root = os.path.abspath(root)
+        self.root = spack.util.url.parse(root)
         self.skip_unstable_versions = skip_unstable_versions
 
     def store(self, fetcher, relative_dest):
         """Fetch and relocate the fetcher's target into our mirror cache."""
 
         # Note this will archive package sources even if they would not
         # normally be cached (e.g. the current tip of an hg/git branch)
-        dst = os.path.join(self.root, relative_dest)
-        mkdirp(os.path.dirname(dst))
+        dst = spack.util.url.join(self.root, relative_dest)
+        local_dst = spack.util.url.local_file_path(dst)
+        if local_dst:
+            mkdirp(os.path.dirname(local_dst))
         fetcher.archive(dst)
```
@scheibelp 

I'm debugging a set of test failures with my changes.  I've narrowed my search down to these particular changes.  I made them without knowing if they're really appropriate, and I would greatly appreciate a generous amount of scrutiny, here.

My main questions concern the `fetcher.archive(dst)` call.  I observe that this call fails in testing (`spack unit-test -k all_mirror`) because the `dst` variable is not a file path, but a URL (due to the changes), and `fetcher.archive()` doesn't work with URLs.  The fetcher is a `VCSFetchStrategy`.

What do you think is the right change to make here?  Given that `dst` is a URL now (that can refer to remote resources), what would a URL-aware version of `fetcher.archive(dst)` look like?  Does `VCSFetchStrategy` also need to be URL-aware?  Do all the fetch strategies need to be?

I worry that either there is a disconnect between the assumptions made by the `MirrorCache` class and the various fetchers, or something is fundamentally wrong with my approach.  Would greatly appreciate your thoughts.

opadron replied:
After some more digging, I see that there is already precedent for fetchers being aware that their `destinations` are (possibly remote) URLs.  See `URLFetchStrategy.archive()`.

I'm pretty sure this is what @robo-wylder was originally trying to point out with his earlier comment.  I reproduce those comments below:

> I'm not sure this is necessary, the fetcher.archive() call itself seems to do the right thing without having to handle remote cases specially. Local directories get created automatically for local paths and for remote paths the fetcher knows how to push the file that already exists locally to the url. Look at the archive() method for the URLFetchStrategy and the implementation of push_to_url() in web.py.

As we can see, however, this is not the case for all fetchers (The `VCSFetchStrategy` still assumes local file path).  This leads me to believe that what needs to be done is to go through all of the fetch strategies and update them to handle the case where their fetch destination is a URL.  Considering that this would be a large change, I think more discussion around it is warranted.

Again, very interested in hearing what you think, @scheibelp.

EDIT: BTW, this "precedent" originated from code contributed by me, so I definitely wouldn't consider it set in stone.

scheibelp replied:
> what would a URL-aware version of fetcher.archive(dst) look like? 

I'm thinking you can copy the `URLFetchStrategy.archive` function's use of `web_util.push_to_url`: Right now all the `VCSFetchStrategy`s need to create their own archive, so I think instead of using `tar czf` with the mirror as the destination, you will want to create a temp dir with the archive, then `web_util.push_to_url` the resulting file in the temp directory.

> Does VCSFetchStrategy also need to be URL-aware? Do all the fetch strategies need to be?

Yes and yes. Overall though I think you should only have to update one function: `VcsFetchStrategy.archive()` - I don't see anything else implementing `archive()`.

If that doesn't clarify what needs to be done (or it seems like I overlooked something) let me know.

Commit:
Update VCSFetchStrategy to work with URLs

Commit:
fix random typo

Commit:
fix testing issue in test/mirror.py

Commit:
update boto MockClientError to derive from the real botocore.ClientError

Commit:
fix isort checks

Commit:
increase coverage

opadron commented:
This is ready for another review.

@scottwittenburg @robo-wylder @scheibelp

alalazo commented on lib/spack/spack/caches.py:
```diff
@@ -59,24 +60,31 @@ def _fetch_cache():
 class MirrorCache(object):
     def __init__(self, root, skip_unstable_versions):
-        self.root = os.path.abspath(root)
+        self.root = spack.util.url.parse(root)
```
I think both the `MirrorCache` class and `__init__` method need a proper docstring. For instance the class was being constructed before with a path-like object (very likely a string?) and now instead wants either a url or a url object. It took a while to figure out why this change is necessary here. 

A docstring explaining the purpose of the `MirrorCache` class at the class level and one explaining what the arguments expected by the constructor are supposed to be would have helped a lot.

opadron replied:
I think @scheibelp would be the best person to add these docstrings.  @scheibelp, feel free to make a PR against my branch.  Alternatively, we can just follow up on a separate Spack PR.

scheibelp replied:
I think the easiest thing to do is I can take a crack at creating a docstring here, and you can add it to the PR.

```
MirrorCache
"""
Keeps track of a Spack mirror that we want to update with new
resources. This includes the URL/path of the mirror as well as
any details about the mirror that describe how we want to 
manage it (e.g. if we only want to include certain types of
resources)

This also provides a `symlink` method for writing cosmetic
(human-readable) symlinks in the mirror, when the mirror
is managed on the local filesystem.
"""

  __init__
  """
  root: a destination to store archived resources. May be a file path (on the local filesystem)
      or a URL. This is used as a target for `FetchStrategy.archive`
  """
```

Having written that, I think there is some overlap between this and the `spack.mirror.Mirror` object (if we decided that excluding unstable resources like branch tips were a property of the mirror itself, then this object could subsume `MirrorCache` entirely), although I think it would make more sense to address that in some other PR.

alalazo commented on lib/spack/spack/caches.py:
```diff
@@ -59,24 +60,31 @@ def _fetch_cache():
         # normally be cached (e.g. the current tip of an hg/git branch)
-        dst = os.path.join(self.root, relative_dest)
-        mkdirp(os.path.dirname(dst))
+        dst = spack.util.url.join(self.root, relative_dest)
+        local_dst = spack.util.url.local_file_path(dst)
+        if local_dst:
+            mkdirp(os.path.dirname(local_dst))
```
This seems like a technical debt we accumulated over time. Why is it the case that to archive locally it's on the caller to ensure `local_dst` exists while to archive remotely (i.e. when `scheme !=' file'`) it's the fetcher that takes care of everything?

opadron replied:
I agree this seems like a debt, but I'm not sure what the best thing to do is.

The root of the difference is the fact that there are no "directories" in `s3://`; everything is stored in a flat hierarchy.  For things like index generation and listing, we do a prefix search.  This is just one example of how key prefixes in S3 are treated like pseudo-folders.

Ideally, all url schemes would get the same treatment, but there's no way to do this without some kind of compromise.  We can't have "folders" everywhere, since there is no such thing with S3, and emulating how S3 treats object keys in `file://` urls would entail potentially `mkdirp`-ing many levels of directories on each write, which raises more questions and usability concerns.  For example, if a user makes a typo in a `file://` url, would it better for Spack to throw an error or blindly create potentially many levels of unintended directories in a user's filesystem?

As another example of how this subtle difference matters, consider two paths: `a/b/c.txt` and `a/b`.  If these were paths from a `file://` url, they couldn't coexist, because `a/b` cannot be both a directory and a file.  For `s3://` urls, this is perfectly fine.

alalazo commented on lib/spack/spack/mirror.py:
```diff
@@ -98,6 +98,63 @@ def from_dict(d, name=None):
         else:
             return Mirror(d['fetch'], d['push'], name)
 
+    @staticmethod
+    def from_args(args):
```
I think an API like this creates an unnecessary coupling between the command line arguments and the `spack.mirror` module, since it's implicitly depending on code in `cmd/mirror.py`. It will also make unit testing slightly harder due to the necessity of mocking argparse namespaces.

Can we extend/refactor instead `Mirror.from_dict`, add a docstring to it and then construct mirrors using `Mirror.from_dict` in commands?

opadron replied:
The purpose here is to construct a valid mirror given a set of user-supplied options, which is something different from what `from_dict` is meant to do.  At the moment, there is only one call site in `cmd/mirror.py` but that will soon change to include other commands that need to accept mirror arguments. (e.g.: #25765) (EDIT: I also remembered that there is already another call site in `cmd/buildcache.py`, which was what originally prompted this PR.)

The main value of the method is the validation and lookup logic, and the intent of factoring it out is to avoid the need for multiple commands to duplicate this logic.  The coupling with spack cmd code is intentional, since that is where the intended call sites would be.

I think the right call is to have `from_args` take keyword arguments instead of an object with attributes, which should simplify unit testing.

alalazo commented on lib/spack/spack/mirror.py:
```diff
@@ -98,6 +98,63 @@ def from_dict(d, name=None):
+                raise ValueError(
+                    '"directory" must be a local path (got a URL)')
+
+            # User meant to provide a path to a local directory.
+            # Ensure that the mirror lookup does not mistake it for a named
+            # mirror.
+            location = 'file://' + location
+            result = spack.mirror.MirrorCollection().lookup(location)
```
This was hard for me to understand and I had to read code carefully to get what's going on. A few suggestions below.

---

I think people would expect the default constructor of a mapping to start empty, like the built-in `dict()`. Here:
```python
spack.mirror.MirrorCollection()
```
reads instead from the global configuration and creates an object with items in it. Besides being unexpected, this default logic couples the class to a global state. Can we extract that logic into a separate function:
```python
def mirror_collection_from_config():
    ...
```
and have the default constructor behave as expected?

---

The `MirrorCollection.lookup` method acts like a factory for mirror items. This is totally unexpected too:
```python
Python 3.8.10, Linux x86_64
>>> import spack.mirror
>>> c = spack.mirror.MirrorCollection()
>>> s = c.lookup('/tmp/mirror')
>>> s
Mirror(fetch_url='/tmp/mirror')
>>> s in c
False
```
Mappings have already a `lookup` method, which is "get" and accepts a default. I think `MirrorCollection.lookup` should be removed in favor of `MirrorCollection.get`.

---

Finally a note that if the class is not meant to be immutable, it should be better to derive `MirrorCollection` from `MutableMapping`.

opadron replied:
These are very valid considerations.  I will follow up with some changes.

opadron replied:
Some points I'm reminded of, having taken a second look at this:

`MirrorCollection` already has a `get()` method that works the same as with other Mappings.  `MirrorCollection.lookup()` works like `get()`, except it has a default return value that depends on the argument passed.  A more verbose name might be `get_or_return_anonymous()`.

Also, `MirrorCollections` *are* immutable.  That's why there is no `__setitem__()` defined.  They're meant to represent a preexisting collection of mirrors that come from a Spack configuration.  This is also why the default constructor doesn't return an empty collection; that would make no sense.  I think the naming should be improved to make it clearer that this class isn't meant to be used as a regular container type.

Perhaps `MirrorConfiguration` or `MirrorGroup` would be a better name?

alalazo replied:
> Also, MirrorCollections are immutable. 

That is fine, but they should be initialized from an iterable of some sort (similarly to what `frozenset` does for instance).

> MirrorCollection already has a get() method that works the same as with other Mappings.

Can't we delete `lookup` and use:
```python
c = spack.mirror.MirrorCollection(iterable)
s = c.get('/tmp/mirror', Mirror(fetch_url='/tmp/mirror'))
```
?

opadron replied:
Yes, I see.  I will look into making those changes.

EDIT: ... except there's already many call sites where this change would need to be duplicated.  The above `get` pattern is going to be a common use case, and so I think we should have some kind of convenience method for it.  Do you think it would be better to duplicate this pattern at every call site?

alalazo commented on lib/spack/spack/mirror.py:
```diff
@@ -98,6 +98,63 @@ def from_dict(d, name=None):
+
+            # User meant to provide the name of a preconfigured mirror.
+            # Ensure that the mirror lookup actually returns a named mirror.
+            result = spack.mirror.MirrorCollection().lookup(location)
```
See comment above

alalazo commented on lib/spack/spack/mirror.py:
```diff
@@ -98,6 +98,63 @@ def from_dict(d, name=None):
+                raise ValueError(
+                    '"{url}" is not a valid URL'.format(url=location))
+
+            result = spack.mirror.MirrorCollection().lookup(location)
```
See comment above

alalazo commented on lib/spack/spack/mirror.py:
```diff
@@ -422,27 +479,25 @@ def create(path, specs, skip_unstable_versions=False):
-        except OSError as e:
-            raise MirrorError(
-                "Cannot create directory '%s':" % mirror_root, str(e))
+    if mirror_root:
+        # Get the absolute path of the root before we start jumping around.
+        if not os.path.isdir(mirror_root):
```
Minor change, but I would rename `mirror_root` to `local_mirror_root` and have a single conditional here:
```python
if local_mirror_root and not os.path.isdir(local_mirror_root):
    ...
```

alalazo commented on lib/spack/spack/s3_handler.py:
```diff
@@ -65,23 +65,7 @@ def s3_open(self, req):
-
-                    raise urllib_error.URLError(err2)
-
+        except (ClientError, ValueError) as err:
```
This is for my understanding of the code. Why do we intercept these exceptions and re-raise them as `URLError`?

opadron replied:
This needs to be done because this method is called as part of `urllib`.  Their API [dictates](https://docs.python.org/3/library/urllib.error.html?highlight=urllib#urllib.error.URLError) that any `urlopen` handler that does not successfully handle a request raises a `URLError`

alalazo commented on lib/spack/spack/test/cmd/mirror.py:
```diff
@@ -48,6 +49,38 @@ def test_regression_8083(tmpdir, capfd, mock_packages, mock_fetch, config):
 
 
+@pytest.mark.regression('17531')
+def test_regression_17531(tmp_scope, tmpdir, capfd, mock_packages, mock_fetch, config):
```
Can we turn this into a parametrized test (it seems it's effectively testing two cases)?

opadron replied:
I can break this up into two separate tests.  I think that makes more sense.

alalazo commented on lib/spack/spack/test/cmd/mirror.py:
```diff
@@ -48,6 +49,38 @@ def test_regression_8083(tmpdir, capfd, mock_packages, mock_fetch, config):
+
+        assert ('Successfully created mirror in ' + tmp_mirror1) in output
+
+        mirror('remove', '--scope', 'command_line', 'named-mirror')
```
Is this done for cleanup?

alalazo commented on lib/spack/spack/test/mirror.py:
```diff
@@ -107,6 +109,59 @@ def test_url_mirror(mock_archive):
     repos.clear()
 
 
+def test_bad_mirror_args():
```
Same request for having a parametrized test.

opadron replied:
I'm not sure how a parameterized version of this test would work.  I can split these into individual tests (should also be simpler given my above suggestion for `from_args()`).

alalazo reviewed (changes requested):
I went through the code and left a few request for changes related to the implementation. Let me know what you think.

Commit:
change Mirror.from_args() to accept kwargs

Commit:
simplify mirror.create()

opadron commented:
@alalazo thanks for the comments.  Please see my responses and questions for further discussion.

Commit:
add from_config() method to MirrorCollection

alalazo commented:
Closing this one as outdated, but anyone here feel free to resurrect it if that seems the right action.
